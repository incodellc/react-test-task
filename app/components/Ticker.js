import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { connectNode } from '../services';

function Ticker(props) {
    const [tickerName, setTickerName] = useState('GOOG');
    const [refreshTime, setRefreshTime] = useState('5000');
    const [disabled, setDisabled] = useState(false);
    const { ticker } = props;

    const submit = (e) => {
        e.preventDefault();
        setDisabled(true);
        connectNode(tickerName, Number(refreshTime));
    };

    const textColor = () => {
        if (ticker.change > 0) return 'green';
        else if (ticker.change < 0) return 'red';
        return 'grey';
    };

    const priceStyle = () => {
        if (Number(refreshTime) <= 5000) {
            return {
                animation: `color${textColor()} ${(
                    Number(refreshTime) / 1000
                ).toString()}s infinite`,
            };
        }
        return { color: textColor() };
    };

    const lastUpdateTime = () => {
        return new Date(Date.parse(ticker.last_trade_time)).toGMTString();
    };

    return (
        <div>
            <form onSubmit={submit} className={'ticker-form'}>
                <select
                    value={refreshTime}
                    onChange={(e) => setRefreshTime(e.target.value)}
                    className={'ticker-select'}
                >
                    <option value={500}>0.5sec</option>
                    <option value={1000}>1sec</option>
                    <option value={5000}>5sec</option>
                    <option value={10000}>10sec</option>
                    <option value={30000}>30sec</option>
                </select>
                <input
                    value={tickerName}
                    onChange={(e) => setTickerName(e.target.value)}
                    placeholder={'Input ticker'}
                    className={'ticker-input'}
                />
                <button
                    className={`ticker-submit ${disabled ? 'disabled' : null}`}
                    disabled={disabled}
                    type={'submit'}
                >
                    Add Ticker
                </button>
            </form>

            <p className={'ticker'}>
                {ticker.ticker} <span style={priceStyle()}>{ticker.price}</span>{' '}
                <span className={'cur'}>USD</span>{' '}
                <span className={'change'} style={{ color: textColor() }}>
                    {ticker.change > 0 ? `+${ticker.change}` : ticker.change} (
                    {ticker.change_percent}%){' '}
                    {textColor() === 'green' ? (
                        <i className={'fas fa-arrow-up'}></i>
                    ) : (<i className={'fas fa-arrow-down'}></i>)}
                </span>
                <span className={'last-update'}>
                    Last Update {ticker.last_trade_time ? lastUpdateTime() : null}
                </span>
            </p>
        </div>
    );
}

const mapStateToProps = (state) => {
    return {
        ticker: state.stockTicker,
    };
};

export default connect(mapStateToProps)(Ticker);

Ticker.propTypes = {
    ticker: PropTypes.object.isRequired,
};
