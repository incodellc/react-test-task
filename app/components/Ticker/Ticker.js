import React from 'react';
import moment from 'moment';
import PropTypes from 'prop-types';

export const Ticker = ({id, item, className}) => {
    const tradeTime = moment( new Date(item.last_trade_time)).format('HH:mm:ss, MM-DD');
    return (
        <tr>
            <th scope="row">{id}</th>
            <td>{item.ticker}</td>
            <td>{item.exchange}</td>
            <td className={`price ${className}`}>{item.price}</td>
            <td>{item.change}</td>
            <td>{item.change_percent}</td>
            <td>{tradeTime}{" "}</td>
            <td>{item.dividend}</td>
            <td>{item.yield}</td>
        </tr>
    );
};

Ticker.propTypes = {
    id: PropTypes.number.isRequired,
    item: PropTypes.object.isRequired,
    className: PropTypes.string.isRequired
};

