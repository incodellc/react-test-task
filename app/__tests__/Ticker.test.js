import React from 'react'
import { shallow } from 'enzyme'
import { Ticker } from '../components/Ticker/Ticker';

describe('Ticker test', () => {
    const props = {
        id: 0,
        item: {
            ticker: 'AAPL',
            exchange: 'NASDAQ',
            price: 116.60,
            change: -0.46,
            change_percent: -0.39,
            last_trade_time: 'Oct 21, 4:00PM EDT',
            dividend: 0.57,
            yield: 1.96
        },
        className: 'arrow-up'
    };

    const ticker = shallow(<Ticker {...props} />);

    it('should renders properly', () => {
        expect(ticker).toMatchSnapshot();
    });

    it('should renders 8 table-data cells (<td>)', () => {
        expect(ticker.find('td')).toHaveLength(8);
    });

    it('should renders 1 table-head cells (<th>)', () => {
        expect(ticker.find('th')).toHaveLength(1);
    });
});

