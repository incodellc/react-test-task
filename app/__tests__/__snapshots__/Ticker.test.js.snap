// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ticker test should renders properly 1`] = `
<tr>
  <th
    scope="row"
  >
    0
  </th>
  <td>
    AAPL
  </td>
  <td>
    NASDAQ
  </td>
  <td
    className="price arrow-up"
  >
    116.6
  </td>
  <td>
    -0.46
  </td>
  <td>
    -0.39
  </td>
  <td>
    Invalid date
     
  </td>
  <td>
    0.57
  </td>
  <td>
    1.96
  </td>
</tr>
`;
